sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -214,18 +251,62 @@
  */
 @Experimental
 public class KafkaIO {
+
+  /**
+   * Get a {@link Coder}. When a non-null coder is provided, use this coder. Otherwise,
+   * attempt to infer a {@link Coder} by extracting the type of the deserialized-class
+   * from the deserializer argument using the {@link Coder} registry.
+   */
+  @Nullable
+  public static <T> Coder<T> getOrInferCoder(CoderRegistry coderRegistry,
+                                             Class<? extends Deserializer<T>> deserializer,
+                                             Coder<T> coder) {
+    if (coder != null) {
+      return coder;
+    }
+
+    if (deserializer == null) {
+      LOG.warn("Could not extract deserializer type from {}", deserializer);
+      return null;
+    }
+
+    // Get Deserializer<T>
+    Type type = deserializer.getGenericInterfaces()[0];
+
+    if (type instanceof ParameterizedType) {
+      // Get T
+      Type actualType = ((ParameterizedType) type).getActualTypeArguments()[0];
+
+      try {
+        @SuppressWarnings("unchecked")
+        Class<T> clazz = (Class<T>) actualType;
+        return coderRegistry.getDefaultCoder(clazz);
+      } catch (CannotProvideCoderException | ClassCastException e) {
+        LOG.warn("Could not infer coder from deserializer type {}: {}",
```

Comment:
I have modified this to let `ClassCastException` be thrown, but `CannotProvideCoderException` is checked and I cannot throw it from `expand` or `validate`. I've modified the logging statement to `LOG.warn(....., e);`,